hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/upgrade/OMPrepareRequest.java

```diff
@@ -0,0 +1,204 @@
+
+      // TODO: Create marker file with txn index.
+
+      LOG.info("OM prepared at log index {}. Returning response {}",
+          ozoneManager.getRatisSnapshotIndex(), omResponse);
+    } catch (IOException e) {
+      response = new OMPrepareResponse(
+          createErrorOMResponse(responseBuilder, e));
+    } catch (InterruptedException e) {
+      response = new OMPrepareResponse(
+          createErrorOMResponse(responseBuilder, new OMException(e,
+              OMException.ResultCodes.INTERNAL_ERROR)));
+    }
+
+    return response;
+  }
+
+  private static void waitForDoubleBufferFlush(
+      OzoneManager ozoneManager, long txnLogIndex)
+      throws InterruptedException, IOException {
+
+    long endTime = System.currentTimeMillis() +
+        DOUBLE_BUFFER_FLUSH_TIMEOUT.toMillis();
+    boolean success = false;
+
+    while (!success && System.currentTimeMillis() < endTime) {
+      // If no transactions have been persisted to the DB, transaction info
+      // will be null, not zero, causing a null pointer exception within
+      // ozoneManager#getRatisSnaphotIndex.
+      // Get the transaction directly instead.
+      OMTransactionInfo txnInfo = ozoneManager.getMetadataManager()
+          .getTransactionInfoTable().get(TRANSACTION_INFO_KEY);
+      if (txnInfo == null) {
+        success = (txnLogIndex == 0);
+      } else {
+        success = (txnInfo.getTransactionIndex() == txnLogIndex);
+      }
+
+      Thread.sleep(DOUBLE_BUFFER_FLUSH_CHECK_INTERVAL.toMillis());
+    }
+
+    // If the timeout waiting for all transactions to reach the state machine
+    // is exceeded, the exception is propagated, resulting in an error response
+    // to the client. They can retry the prepare request.
+    if (!success) {
+      throw new IOException(String.format("After waiting for %d seconds, " +
+              "State Machine has not applied  all the transactions.",
+          DOUBLE_BUFFER_FLUSH_TIMEOUT.toMillis() * 1000));
+    }
+  }
+
+  /**
+   * Take a snapshot of the state machine at the last index, and purge ALL logs.
+   * @param impl RaftServerImpl instance
+   * @throws IOException on Error.
+   */
+  public static long takeSnapshotAndPurgeLogs(RaftServerImpl impl)
+      throws IOException {
+
+    StateMachine stateMachine = impl.getStateMachine();
+    long snapshotIndex = stateMachine.takeSnapshot();
+
+    // If the snapshot indices from Ratis and the state machine do not match,
+    // the exception is propagated, resulting in an error response to the
+    // client. They can retry the prepare request.
+    if (snapshotIndex != stateMachine.getLastAppliedTermIndex().getIndex()) {
+      throw new IOException("Index from Snapshot does not match last applied " +
+          "Index");
+    }
+
+    RaftLog raftLog = impl.getState().getLog();
+    // In order to get rid of all logs, make sure we also account for
+    // intermediate Ratis entries that do not pertain to OM.
+    long lastIndex = Math.max(snapshotIndex,
```

Comment:
This was originally done defensively against meta transactions that Ratis may put in the log, which are not related to OM activities. @avijayanhwx and I are unable to reproduce earlier errors we thought we saw related to this, so we will remove this max statement and just use snapshotIndex. If an error surfaces later under more robust prepare and upgrade testing, we will re-address this issue, but for now we can remove it to avoid unintended side effects this might cause.